tests/harness.rs

```diff
@@ -0,0 +1,307 @@
+        })
+    }
+
+    /// Add a signed transaction from a provided signer to the sim cache.
+    pub fn add_tx(&self, signer: &PrivateKeySigner, nonce: u64, value: U256, mpfpg: u128) {
+        let tx = new_signed_tx_with_max_fee(signer, nonce, value, mpfpg, TEST_TX_MAX_FEE)
+            .expect("tx signing");
+        // group index 0 for simplicity in tests
+        self.simulator.sim_cache.add_tx(tx, 0);
+    }
+
+    /// Mine additional blocks on the underlying Anvil instance and update the sim_env with the latest headers.
+    pub async fn mine_blocks(&self, count: u64) -> eyre::Result<()> {
+        if count == 0 {
+            return Ok(());
+        }
+
+        self.host.anvil_mine(Some(count), Some(1)).await?;
+        self.rollup.anvil_mine(Some(count), Some(1)).await?;
+
+        let (ru, host) = self.get_headers().await?;
+        self.tick_from_headers(ru, host).await;
+
+        Ok(())
+    }
+
+    /// Starts the simulator task.
+    pub async fn start(&mut self) {
+        if self.simulator_handle.is_some() {
+            tracing::warn!("TestHarness simulator already running");
+            return;
+        }
+        tracing::debug!("TestHarness starting simulator task");
+
+        // Spawn the simulator background task
+        let cache = self.simulator.sim_cache.clone();
+
+        // Rollup provider
+        let ru_provider = RootProvider::<Ethereum>::new_http(self.rollup.anvil().endpoint_url());
+
+        // Host provider
+        let host_provider = self.host_provider().await;
+
+        // Wire up the simulator with the providers and submit channels
+        let submit_tx = self.submit_tx.clone();
+        let simulator = Simulator::new(
+            &self.config,
+            host_provider,
+            ru_provider,
+            self.simulator.sim_env_rx.clone(),
+        );
+
+        // Keep the JoinHandle on the harness so the task isn't aborted when
+        // this function returns.
+        let jh = simulator.spawn_simulator_task(cache, submit_tx);
+        self.simulator_handle = Some(jh);
+        tracing::debug!("TestHarness spawned simulator task");
+    }
+
+    /// Returns the latest rollup and host headers.
+    pub async fn get_headers(&self) -> eyre::Result<Headers> {
+        let ru_block = self.rollup.get_block(BlockId::latest()).await?;
+        let ru_header = ru_block.expect("rollup latest exists").header.inner;
+
+        let host_block = self.host.get_block(BlockId::latest()).await?;
+        let host_header = host_block.expect("host latest exists").header.inner;
+
+        tracing::debug!(
+            rollup_header_number = ru_header.number,
+            rollup_header_gas_limit = ru_header.gas_limit
+        );
+
+        let headers = (ru_header, host_header);
+        Ok(headers)
+    }
+
+    /// Tick a new `SimEnv` computed from the current latest rollup and host headers.
+    pub async fn tick_from_headers(&self, prev_ru_header: Header, prev_host_header: Header) {
```

Comment:
this is now duplicating env construction code from the env task. is there a reason we're not using an EnvTask here? it seems like we could easily have one running but control anvil to only produce WS notifications when a `tick` function is called